webapp/src/Controller/API/SubmissionController.php

```diff
@@ -219,23 +318,120 @@ public function addSubmissionAction(Request $request)
             $entryPoint = $request->request->get('entry_point');
         }
 
-        // Get the files we want to submit
-        $files = $request->files->get('code') ?: [];
-        if (!is_array($files)) {
-            $files = [$files];
+        $time = null;
+        if ($timeString = $request->request->get('time')) {
+            if ($this->isGranted('ROLE_API_WRITER')) {
+                try {
+                    $time = Utils::toEpochFloat($timeString);
+                } catch (Exception $e) {
+                    throw new BadRequestHttpException(sprintf('Can not parse time %s', $timeString));
+                }
+            } else {
+                throw new BadRequestHttpException('A team can not assign time');
+            }
+        }
+
+        if ($submissionId = $request->request->get('id')) {
+            if ($this->isGranted('ROLE_API_WRITER')) {
+                // Check if we already have a submission with this ID
+                $existingSubmission = $this->em->createQueryBuilder()
+                    ->from(Submission::class, 's')
+                    ->select('s')
+                    ->andWhere('(s.externalid IS NULL AND s.submitid = :submitid) OR s.externalid = :submitid')
+                    ->andWhere('s.contest = :contest')
+                    ->setParameter(':submitid', $submissionId)
+                    ->setParameter(':contest', $problem->getContest())
+                    ->getQuery()
+                    ->getOneOrNullResult();
+                if ($existingSubmission !== null) {
+                    throw new BadRequestHttpException(sprintf("Submission with ID %s already exists", $submissionId));
+                }
+            } else {
+                throw new BadRequestHttpException('A team can not assign id');
+            }
+        }
+
+        $tempFiles = [];
+
+        if ($request->request->has('files')) {
+            // CCS spec format, files are a ZIP, get them and transform them into a file object
+            $filesList = $request->request->get('files');
+            if (!is_array($filesList) || count($filesList) !== 1 || !isset($filesList[0]['data'])) {
+                throw new BadRequestHttpException("The 'files' attribute should be an array with a single item, containing an object with a base64 encoded data field");
+            }
+
+            if (isset($filesList[0]['mime']) && $filesList[0]['mime'] !== 'application/zip') {
+                throw new BadRequestHttpException("The 'files[0].mime' attribute should be application/zip if provided");
+            }
+
+            $data        = $filesList[0]['data'];
+            $decodedData = base64_decode($data);
```

Comment:
Do we want to set the strict flag? Seems good to be strict I think.

Reply:
Yep sounds good.